### ISSUE TITLE
Improving Object-Oriented design

### ISSUE THREAD

**Issue by mtymek** [serbanghita/Mobile-Detect#410]
Hi Serban,

As discussed over e-mail, I'd like to trigger a brainstorm on how to improve OO design of this library with new major release. I have a proposal that should produce something simple, easy to use and maintainable.

First of all, I'd like to start with the main class. It should serve one purpose: converting HTTP headers into useful information about user device. Current implementation has 29 public methods, for doing variety of things:

```php
Class Mobile_Detect
{
    public static function getScriptVersion();
    public function setHttpHeaders($httpHeaders = null);
    public function getHttpHeaders();
    public function getHttpHeader($header);
    public function getMobileHeaders();
    public function getUaHttpHeaders();
    public function setUserAgent($userAgent = null);
    public function getUserAgent();
    public function setDetectionType($type = null);
    public function getMatchingRegex();
    public function getMatchesArray();
    public static function getPhoneDevices();
    public static function getTabletDevices();
    public static function getUserAgents();
    public static function getBrowsers();
    public static function getUtilities();
    public static function getMobileDetectionRules();
    public function getMobileDetectionRulesExtended();
    public function getRules();
    public static function getOperatingSystems();
    public function checkHttpHeadersForMobile();
    public function isMobile($userAgent = null, $httpHeaders = null);
    public function isTablet($userAgent = null, $httpHeaders = null);
    public function is($key, $userAgent = null, $httpHeaders = null);
    public function match($regex, $userAgent = null);
    public static function getProperties();
    public function prepareVersionNo($ver);
    public function version($propertyName, $type = self::VERSION_TYPE_STRING);
    public function mobileGrade();
}
```

It has following main responsibilities:
1. parsing HTTP headers from `$_SERVER`
2. providing dictionaries (`getPhoneDevices()`, `getTabletDevices()`, etc)
3. detecting device type and parameters
4. providing handy interface for testing detected device type and parameters (`isMobile()`, `isTablet()`...)

Let's separate them:

1. HTTP headers can be handled by any PSR-7 compatible library, so we can strip this out.
2. If dictionaries are actually used outside, they can be placed in separate classes. For main class, it should be implementation detail, not accessible from outside.
3. Detecting should stay in main class. Exactly as in your `devel-3-refactoring` branch, `detect()` method should return an object responsible for holding device information...
4. ...which should come with handy interface.

I would like to propose something very simple for `MobileDetect` contract:

```php
use Psr\Http\Message\RequestInterface;

interface MobileDetectInterface
{
    /**
     * @param RequestInterface $request
     * @return DeviceInterface
     */
    public function detect(RequestInterface $request);
}
```

Implementation would relay on PSR-7 library for parsing headers - you would be able to use canonical names directly:

```php
class MobileDetect implements MobileDetectInterface
{
    /**
     * @param RequestInterface $request
     * @return DeviceInterface
     */
    public function detect(RequestInterface $request)
    {
	// real detection engine would go here
	if (preg_match($request->getHeaderLine('User-Agent'), '/Android/')) {
	    return new Device('Android');
	}
	// ...
    }    
}
```

Unit testing such class would be simpler too. I would argue that single unit test with good data provider would be enough:

```php
class MobileDetectTest extends PHPUnit_Framework_TestCase
{
    public function provideDataForDetect()
    {
        return [
            'iphone_with_ios_50' => [
                'headers' => [
                    'User-Agent' => 'Mozilla/5.0 (iPhone; CPU iPhone OS 5_0 like Mac OS X) AppleWebKit/534.46 (KHTML, like Gecko) Version/5.1 Mobile/9A334 Safari/7534.48.3',		    
                ],
                'isPhone' => true,
                'isTablet' => false,
            ],
	    // ...
        ];
    }

    /**
     * @dataProvider provideDataForDetect
     */
    public function testDetect($headers, $isPhone, $isTablet /*, ...*/)
    {
        $request = new Request(null, null, 'php://temp', $headers);
        $mobileDetect = new MobileDetect();
        $device = $mobileDetect->detect($request);
        $this->assertEquals($isPhone, $device->isPhone());
        $this->assertEquals($isTablet, $device->isTablet());
    }
}
```

I would start the work by removing header processing, updating class to use `PSR-7`, then working on `detect()` method and `Device` class.
On top of that, after getting test suite to work, you would be able to refactor the actual detection engine. It would become implementation detail of `MobileDetect` class, so no need to inject it, or create factory for it. 

Thoughts?

**Comment by nicktacular** [comment]
Have you taken a look at the re-designed classes in version 3?

**Comment by mtymek** [comment]
Hi Nick,
Yes, Iooked - that's why I mentioned `devel-3-refactoring` branch. That's from where I took the idea of `detect()` method returning `Device` object. What I don't like there though is how complex it is - from custom header parsing to how `MobileDetectFactory` works. That's why I proposed different approach, to start with something simpler, better from OOP design (BTW I'm also willing to help with coding at some point:) ).
So, what is your opinion on what I proposed?

**Comment by wheelsandcogs** [comment]
Hi, is there an ETA / roadmap for the refactored v3? I realise you may be too busy, but a rough timescale would help me decide if I need to fork or not.

**Comment by nicktacular** [comment]
@wheelsandcogs we are currently working on a finite list of items that we need to complete prior to our initial 3.0 release. That list is relatively small but it's difficult to give an exact timeline at this point, but I would say that this will be happening soon. I will discuss with @serbanghita and we'll post in here when we have a better idea of the remaining items.

**Comment by nicktacular** [comment]
@mtymek:

> HTTP headers can be handled by any PSR-7 compatible library

Yes, however, we still need to provide the ability to extract from the PHP environment using `$_SERVER` superglobal, so that needs to be maintained since this library can be used without a PSR-7 compatible library.

> If dictionaries are actually used outside, they can be placed in separate classes

The device libraries have been extracted into the `MobileDetect\Providers` namespace.

> Detecting should stay in main class. Exactly as in your `devel-3-refactoring branch`, `detect()` method should return an object responsible for holding device information

The `detect` method does return a `MobileDetect\Device\Device` object which follows the `MobileDetect\Device\DeviceInterface`. Optionally, passing a classname as a string to the method will instantiate using the factory method in `MobileDetect\Device\DeviceInterface::create`, returning an instance of that interface with whichever implementation a user might want. Example `$mobileDetect->detect('\\MyNamespace\\MyDeviceClass')` will create an instance of the `\MyNamespace\MyDeviceClass` class assuming it implements the interface.

> ...which should come with handy interface.

Yup.

> I would like to propose something very simple for MobileDetect contract: `public function detect(RequestInterface $request);`

Headers should be injected at instantiation or using a setter. Injecting at this time means a user must always utilize the `RequestInterface` at runtime versus at bootstrapping. Bootstrapping frequently makes more sense which is why we've chosen not to do this. As far as the interface is concerned, I'm not sure it's necessary for the main detection class. Can you explain further why this would make sense?

> Implementation would relay on PSR-7 library for parsing headers - you would be able to use canonical names directly

Yes, that's true, but not possible because we can't assume that the Mobile Detect users all have a PSR-7 compatible HTTP lib that they're using. We have implemented the ability to optionally provide the headers utilizing a PSR-7 compatible lib if a user wishes to do so. See 4ef3732096b5f25085d0aab65d38c2c993889ef1 / #409.

I will take a look at what we can do with the header processing, which is unavoidable given our need to only optionally suggest PSR-7 and not actually require it. I will update this thread when I've taken a look at what can be done here.

**Comment by serbanghita** [comment]
I restarted the research work and I'm working on finishing the Integration Tests on `devel-3` https://github.com/serbanghita/Mobile-Detect/tree/devel-3

Nick and I will discuss a public roadmap next week probably, after I have a big suite of tests finished.
